docs: remove duplicate documentation (#262)
0xPolygonMiden/miden-client: README.md

The client's main responsibility is to maintain a partial view of the blockchain which allows for locally executing and proving transactions. It keeps a local store of various entities that periodically get updated by syncing with the node.

## Dependencies

### Required

- [Rust installation](https://www.rust-lang.org/tools/install) with version 1.77 or higher.

### Optional

- [Cargo Make](https://github.com/sagiegurari/cargo-make)

## Usage

### Adding miden-client as a dependency

In order to utilize the `miden-client` library, you can add the dependency to your project's `Cargo.toml` file:

````toml
miden-client = { version = "0.2" }
````

### Running miden-client from CLI

You can either build from source with:

```bash
cargo build --release
```

Once the binary is built, you can find it on `./target/release/miden-client`.

Or you can install the CLI from crates-io with:

```bash
cargo install miden-client
```

Check the official documentation for more details reagarding:

> - [Features](https://docs.polygon.technology/miden/miden-client/library/#features), such us faster PoW for account creation and lower PoW difficulty for development.

Reviewer: ```suggestion
- [Library Features](https://docs.polygon.technology/miden/miden-client/library/#features)
```

Reviewer: I think saying that saying "lower PoW difficulty" is the most precise way to describe the `testing` feature but overall I think I'd prefer to remove the description from here

Reviewer: Also, should we create a smaller section here to mention features? Maybe at the level of this sub-section. Since they are pretty important we might want to highlight them more

Author: I thought the same but was unsure whether to add it or not to avoid duplicate documentation. However, features will likely be something that won't change much so it's not much of a big deal

- [Configuration](https://docs.polygon.technology/miden/miden-client/cli-config/)
- [Examples](https://0xpolygonmiden.github.io/miden-base/introduction/getting-started.html)

> - [CLI Reference](https://docs.polygon.technology/miden/miden-client/cli-reference/#types-of-transaction)

Reviewer: Should these point to the repo's files instead of the generated docs URL?

Author: hmmm it's probably better. I'd keep one link to the generated docs URL as well.

Reviewer 2: However, if this is the README, I would recommend linking to the repo files, and then adding a single link to the docs site somewhere ... but up to you :)

The client's main responsibility is to maintain a partial view of the blockchain which allows for locally executing and proving transactions. It keeps a local store of various entities that periodically get updated by syncing with the node.

> ## Dependencies

Reviewer 3: I'm not sure this section is needed.

- We can mention `cargo-make` dependency in the `Makefile` section below.
- The min `rustc` version is mentioned in the badge in the latest readme (and we can probably link this badge to the Rust install instructions). But we can also make an explicit reference to this in the first sentence in the `Usage` section.

Author: Oh, I also forgot to add sqlite in the dependency section. If we remove the section, where should I mention sqlite?

Reviewer 3: Maybe this could go into the intro to the `Usage` section (something similar to how we did it for [miden-node](https://github.com/0xPolygonMiden/miden-node?tab=readme-ov-file#usage)).

Author: done! I'm wondering whether we should mention that it should be possible to ditch sqlite3 and use another db when you provide its own `Store` implementation (I think we'd need to put the sqlite3 related crates under a feature flag) . What do you think? Maybe that should be part of the documentation and not the readme though

Reviewer 3: Ah! Good point re `sqlite`. I think it is required only for the CLI - so, if we do mention it here, it should go into the CLI sub-section.

In the library API docs, we should also mention that we provide an SQLite-based `Store` implementation but users are free to provide custom implementations (I'm not sure we need a feature flag for this) - but that's probably for a different PR.

Author: I think currently rusqlite comes in both the CLI and the library. We use it to run integration tests for example. It's sort of our "default" store implementation. That's why I find it reasonable to put it under a sqlite feature flag. Perhaps it's better to make an issue to address this in the documentation as well as putting the sqlite store under a feature flag. What do you think?

Reviewer 3: > Perhaps it's better to make an issue to address this in the documentation as well as putting the sqlite store under a feature flag. What do you think?

Agreed!

Author: created #284

### Required

- [Rust installation](https://www.rust-lang.org/tools/install) with version 1.77 or higher.

### Optional

- [Cargo Make](https://github.com/sagiegurari/cargo-make)

## Usage

### Adding miden-client as a dependency

In order to utilize the `miden-client` library, you can add the dependency to your project's `Cargo.toml` file:

````toml
miden-client = { version = "0.2" }
````

#### Features

- `concurrent`: used to enable concurrent proofs generation
- `testing`: useful feature that lowers PoW difficulty when enabled. Only use this during development and not on production.
- `test_utils`: used on unit tests to use the mocked rpc api 
- `integration`: only used to run integration tests and separate them from unit tests

concurrent = [
    "miden-lib/concurrent",
    "miden-objects/concurrent",
    "miden-tx/concurrent",
]
default = ["std"]
integration = ["testing", "concurrent", "uuid"]
std = ["miden-objects/std"]
testing = ["miden-objects/testing", "miden-lib/testing"]

> test_utils = ["miden-objects/testing"]

Reviewer 3: Do we need to include this part in the readme?

Author: oh no, my bad. I copied it from cargo toml to make the features list and forgot to delete it.

### Running `miden-client`'s CLI

You can either build from source with:

```bash
cargo build --release
```

Once the binary is built, you can find it on `./target/release/miden-client`.

Or you can install the CLI from crates-io with:

```bash
cargo install miden-client
```

Check the official documentation for more details reagarding:

- [Library Features](./docs/library.md#features)
- [Configuration](./docs/cli-config.md)
- [Examples](https://0xpolygonmiden.github.io/miden-base/introduction/getting-started.html)
- [CLI Reference](./docs/cli-reference.md#types-of-transaction)

> - [Online Documentation](https://docs.polygon.technology/miden/miden-client)

Reviewer 3: I'm wondering if we should move these links up to the `Overview` section. Also:

1. Should all links be absolute and point to doc sites?
2. Ideally, the link for library API should go to the [docs.rs site](https://docs.rs/crate/miden-client/0.2.0) (once we have it fixed).
3. I would probably reduce the number of links here to something like:
  a. `Get started` which would be the link to the tutorial.
  b. `CLI reference` which should include config reference.
  c. `Library API` which should be the link to the docs.rs site.

Author: I agree on moving those up to the overview section. Also:

1. I think either is fine. I have a slight preference over having relative paths to local documentation, while keeping one link to the docs site.
2. I think there's no Library API one, but I can add one and remove the Library Features one.
3.
  a. ok
  b. Should we move the contents of `docs/cli-config.md` into `docs/cli-reference`? Or did you mean to add it as a subitem
  c. ok

Author: oh, regarding the docs.rs site, isn't it usual to add the link in the README's header? Here's an example from [macroquad's repo](https://github.com/not-fl3/macroquad)

![image](https://github.com/0xPolygonMiden/miden-client/assets/25107475/63e09f78-2f97-40b9-9d5d-0b663fd341fc)

Reviewer 3: > oh, regarding the docs.rs site, isn't it usual to add the link in the README's header?

I think since we have more than just library API docs, having just one link in the header may be confusing. If we do have that link (in addition to the more detailed description of the docs), I'd probably point it to the Polygon docs site.

> Should we move the contents of `docs/cli-config.md` into `docs/cli-reference`? Or did you mean to add it as a subitem

Ideally, there should be a single `Miden CLI` section with config and reference as sub-sections and the link should be to `Miden CLI` section (I think this is largely how it works now, except we don't have a way to link to `Miden CLI` section).

Author: > I think since we have more than just library API docs, having just one link in the header may be confusing. If we do have that link (in addition to the more detailed description of the docs), I'd probably point it to the Polygon docs site.

I added the docs.rs badge and removed the Library API link (the link to the documentation is under "Online Documentation"). Do you prefer "Library API" over that one?

Reviewer 3: > I added the docs.rs badge and removed the Library API link (the link to the documentation is under "Online Documentation"). Do you prefer "Library API" over that one?

I'd prefer to have it as it was (without the docs badge) as I would expect the badge to point to the docs to the entire client, not one aspect of it. I maybe for now we can skip `Library API` link since it is not working anyway - but I would like to revisit the structure of the client docs again in the future (I think some sections may be misplaced). Let's create an issue for this.

Reviewer 3: Also, for badges - could you rebase this from the latest code. I think we do badges differently there.

Author: > Also, for badges - could you rebase this from the latest code. I think we do badges differently there.

I think there are some commits in `main` that were never merged into `next`. Should we make a PR from `main` to `next` and then rebase our branches into main?

Reviewer 3: > Should we make a PR from `main` to `next` and then rebase our branches into main?

Yes - let's do this.

Author: rebased!

### Makefile

As mentioned before, we use `cargo-make` to encapsulate some tasks, such as running lints and tests. You can check out [Makefile.toml](./Makefile.toml) for all possible tasks to run.

## Testing

To test the project's code, we provide both unit tests (which can be run with `cargo test`) and integration tests. For more info on integration tests, refer to the [integration testing document](./tests/README.md)

## Changelog

> You can keep track of the client's changes [here](./CHANGELOG.md).

Reviewer 3: I'm not sure we need this section.

Author: I'll remove it then

The client's main responsibility is to maintain a partial view of the blockchain which allows for locally executing and proving transactions. It keeps a local store of various entities that periodically get updated by syncing with the node.

For more info check:

- [Get started](https://0xpolygonmiden.github.io/miden-base/introduction/getting-started.html)
- [CLI Reference](./docs/cli-reference.md#types-of-transaction)
    - [Configuration](./docs/cli-config.md)
- [Online Documentation](https://docs.polygon.technology/miden/miden-client)

## Usage

Before you can use the Miden client, you'll need to make sure you have both
[Rust](https://www.rust-lang.org/tools/install) and sqlite3 installed. Miden

> client v0.2 requires rust version 1.77 or higher.

Reviewer: ```suggestion
client v0.2 requires rust version **1.77** or higher.
```

### Adding miden-client as a dependency

In order to utilize the `miden-client` library, you can add the dependency to your project's `Cargo.toml` file:

````toml
miden-client = { version = "0.2" }
````

#### Features

- `concurrent`: used to enable concurrent proofs generation
- `testing`: useful feature that lowers PoW difficulty when enabled. Only use this during development and not on production.

> - `test_utils`: used on unit tests to use the mocked rpc api 

Reviewer: ```suggestion
- `test_utils`: used on unit tests to use the mocked RPC API.
```

> - `integration`: only used to run integration tests and separate them from unit tests

Reviewer: Should we separate these to a sub-section? The reason being that they are not really relevant to the end user and it could be confusing to keep them

Author: I could move them into the `Testing` subsection that is below on the readme.

Reviewer: Yes, I think that would be nice

### Running `miden-client`'s CLI

You can either build from source with:

```bash
cargo build --release
```

Once the binary is built, you can find it on `./target/release/miden-client`.

Or you can install the CLI from crates-io with:

```bash
cargo install miden-client
```

### Makefile

> As mentioned before, we use [cargo-make](https://github.com/sagiegurari/cargo-make) to encapsulate some tasks, such as running lints and tests. You can check out [Makefile.toml](./Makefile.toml) for all possible tasks to run.

Reviewer: ```suggestion
As mentioned before, we use [cargo-make](https://github.com/sagiegurari/cargo-make) to encapsulate some tasks, such as running lints and tests. You can check out [Makefile.toml](./Makefile.toml) for all available tasks.
```

## Testing

To test the project's code, we provide both unit tests (which can be run with `cargo test`) and integration tests. For more info on integration tests, refer to the [integration testing document](./tests/README.md)

## Contributing

> Interested in contributing with the project? Check [CONTRIBUTING.md](./CONTRIBUTING.md).

Reviewer: ```suggestion
Interested in contributing? Check [CONTRIBUTING.md](./CONTRIBUTING.md).
```
